components/calendar/src/chinese_based.rs

```diff
@@ -0,0 +1,346 @@
+        if lhs_argument == 12
+            && (Self::no_major_solar_term(month_after_eleventh)
+                || Self::no_major_solar_term(month_after_twelfth))
+        {
+            Self::new_moon_on_or_after((month_after_twelfth + 1).as_moment())
+        } else {
+            month_after_twelfth
+        }
+    }
+
+    /// Get the moment of the nearest winter solstice on or before a given fixed date
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5359-L5368
+    pub(crate) fn winter_solstice_on_or_before(date: RataDie) -> RataDie {
+        let approx = Astronomical::estimate_prior_solar_longitude(
+            270.0,
+            Self::midnight((date + 1).as_moment()),
+        );
+        let mut iters = 0;
+        let max_iters = 367;
+        let mut day = Moment::new(libm::floor(approx.inner() - 1.0));
+        while iters < max_iters && 270.0 >= Astronomical::solar_longitude(Self::midnight(day + 1.0))
+        {
+            iters += 1;
+            day += 1.0;
+        }
+        debug_assert!(
+            iters < max_iters,
+            "Number of iterations was higher than expected"
+        );
+        day.as_rata_die()
+    }
+
+    /// Get the fixed date of the nearest Lunar New Year on or before a given fixed date.
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5396-L5405
+    pub(crate) fn new_year_on_or_before_fixed_date(date: RataDie) -> RataDie {
+        let new_year = Self::new_year_in_sui(date);
+        if date >= new_year {
+            new_year
+        } else {
+            Self::new_year_in_sui(date - 180)
+        }
+    }
+
+    /// Get a Date<C> from a fixed date
+    ///
+    /// Months are calculated by iterating through the dates of new moons until finding the last month which
+    /// does not exceed the given fixed date. The day of month is calculated by subtracting the fixed date
+    /// from the fixed date of the beginning of the month.
+    ///
+    /// The calculation for `elapsed_years` in this function is based on code from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5414-L5459
+    pub(crate) fn chinese_based_date_from_fixed(date: RataDie) -> Date<C> {
+        let new_year = Self::new_year_on_or_before_fixed_date(date);
+        let elapsed_years =
+            libm::floor(1.5 - 1.0 / 12.0 + ((new_year - C::EPOCH) as f64) / MEAN_TROPICAL_YEAR);
+        let elapsed_years_int = i64_to_i32(elapsed_years as i64);
+        debug_assert!(
+            matches!(elapsed_years_int, I32Result::WithinRange(_)),
+            "Year should be in range of i32"
+        );
+        let year = elapsed_years_int.saturate();
```

Comment:
Nit: We should saturate the date (ok for a follow up PR)

Reply:
Leaving this for follow-up #3662